java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java

```diff
@@ -152,130 +130,380 @@ private void execute(String signature, SquidClassLoader classLoader) {
-        // TODO use constraintManager.createMethodSymbolicValue to create relational SV for equals
-        SymbolicValue returnSV = constraintManager.createSymbolicValue(instruction);
-        if (isStatic) {
-          // follow only static invocations for now.
-          String signature = instruction.fieldOrMethod.completeSignature();
-          MethodBehavior methodInvokedBehavior = behaviorCache.get(signature);
-          if (methodInvokedBehavior != null && methodInvokedBehavior.isComplete()) {
-            methodInvokedBehavior
-              .happyPathYields()
-              .forEach(yield ->
-                yield.statesAfterInvocation(Lists.reverse(pop.values), Collections.emptyList(), pop.state, () -> returnSV).forEach(ps -> {
-                  checkerDispatcher.methodYield = yield;
-                  checkerDispatcher.addTransition(ps);
-                  checkerDispatcher.methodYield = null;
-                }));
-            methodInvokedBehavior
-              .exceptionalPathYields()
-              .forEach(yield ->
-                yield.statesAfterInvocation(
-                  Lists.reverse(pop.values), Collections.emptyList(), pop.state, () -> constraintManager.createExceptionalSymbolicValue(yield.exceptionType())).forEach(ps -> {
-                  ps.storeExitValue();
-                  enqueue(new ProgramPoint(exitBlock), ps);
-                }));
-            return;
-          }
-
-        }
-        if (instruction.hasReturnValue()) {
-          programState = pop.state;
-        } else {
-          programState = pop.state.stackValue(returnSV);
+      case POP2:
+        programState = programState.unstackValue(2).state;
+        break;
+      case DUP:
+        sv = programState.peekValue();
+        Preconditions.checkNotNull(sv, "DUP on empty stack");
+        programState = programState.stackValue(sv);
+        break;
+      case DUP_X1:
+        pop = programState.unstackValue(2);
+        Preconditions.checkState(pop.values.size() == 2, "DUP_X1 needs 2 values on stack");
+        programState = pop.state.stackValue(pop.values.get(0)).stackValue(pop.values.get(1)).stackValue(pop.values.get(0));
+        break;
+      case DUP_X2:
+        pop = programState.unstackValue(3);
+        Preconditions.checkState(pop.values.size() == 3, "DUP_X2 needs 3 values on stack");
+        programState = pop.state.stackValue(pop.values.get(0)).stackValue(pop.values.get(2)).stackValue(pop.values.get(1)).stackValue(pop.values.get(0));
+        break;
+      case DUP2:
+        pop = programState.unstackValue(2);
+        Preconditions.checkState(pop.values.size() == 2, "DUP2 needs 2 values on stack");
+        programState = pop.state.stackValue(pop.values.get(1)).stackValue(pop.values.get(0)).stackValue(pop.values.get(1)).stackValue(pop.values.get(0));
+        break;
+      case DUP2_X1:
+        pop = programState.unstackValue(3);
+        Preconditions.checkState(pop.values.size() == 3, "DUP2_X1 needs 3 values on stack");
```

Comment:
I have the feeling that the part with 
```
  pop = programState.unstackValue(3);
  Preconditions.checkState(pop.values.size() == 3, "DUP2_X1 needs 3 values on stack");
``` 
should probably be extracted into a method doing the assertion with a generated message to simplify a bit the switch case code.